storage/database.go

```diff
@@ -158,34 +172,220 @@ func NewDatabase(
+	if err := d.addNamespacesWithLock(adds); err != nil {
+		enrichedErr := fmt.Errorf("unable to add namespaces: %v", err)
+		d.log.Errorf("%v", enrichedErr)
+		return err
+	}
+
+	// TODO(prateek): create issue for this ~ namepace updates need to be handled better,
+	// the current implementation updates namespaces by closing and re-creating them.
+	// There are a few problems with this approach:
+	// (1) it's too expensive to flush all the data, and re-bootstrap it for a namespace
+	// (2) we stop accepting writes during the period between when a namespace is removed
+	// until it's added back. This is worse when you consider a KV update can propagate
+	// to all nodes at the same time.
+	// (3): we are not controlling bg processing (repairs/bootstrapping/etc) whilst updating
+	// active namespaces.
+
+	// log that updates and removals are skipped
+	if len(removes) > 0 || len(updates) > 0 {
+		d.log.Warnf("Skipping namespace removals and updates, restart process if you want changes to take effect.")
+	}
+
+	// enqueue bootstraps if new namespaces
+	if len(adds) > 0 {
+		d.queueBootstrapWithLock()
+	}
+
+	return nil
+}
+
+func (d *db) namespaceDeltaWithLock(newNamespaces namespace.Map) ([]ts.ID, []namespace.Metadata, []namespace.Metadata) {
+	var (
+		existing = d.namespaces
+		removes  []ts.ID
+		adds     []namespace.Metadata
+		updates  []namespace.Metadata
+	)
+
+	// check if existing namespaces exist in newNamespaces
+	for _, ns := range existing {
+		newMd, err := newNamespaces.Get(ns.ID())
+
+		// if a namespace doesn't exist in newNamespaces, mark for removal
+		if err != nil {
+			removes = append(removes, ns.ID())
+			continue
+		}
+
+		// if namespace exists in newNamespaces, check if options are the same
+		optionsSame := newMd.Options().Equal(ns.Options())
+
+		// if options are the same, we don't need to do anything
+		if optionsSame {
+			continue
+		}
+
+		// if options are not the same, we mark for updates
+		updates = append(updates, newMd)
+	}
+
+	// check for any namespaces that need to be added
+	for _, ns := range newNamespaces.Metadatas() {
+		_, exists := d.namespaces[ns.ID().Hash()]
+		if !exists {
+			adds = append(adds, ns)
+		}
+	}
+
+	return removes, adds, updates
+}
+
+func (d *db) logNamespaceUpdate(removes []ts.ID, adds, updates []namespace.Metadata) error {
+	removalString, err := tsIDs(removes).String()
+	if err != nil {
+		return fmt.Errorf("unable to format removal, err = %v", err)
+	}
+
+	addString, err := metadatas(adds).String()
+	if err != nil {
+		return fmt.Errorf("unable to format adds, err = %v", err)
+	}
+
+	updateString, err := metadatas(updates).String()
+	if err != nil {
+		return fmt.Errorf("unable to format updates, err = %v", err)
+	}
+
+	// log scheduled operation
+	d.log.WithFields(
+		xlog.NewLogField("removals", removalString),
+		xlog.NewLogField("adds", addString),
+		xlog.NewLogField("updates", updateString),
+	).Infof("updating database namespaces")
```

Comment:
Sure thing, I more meant in the code but yeah a log message should inform the reader of the code too I suppose.